core/ir/aarch64/codec.txt

```diff
@@ -1097,20 +1104,23 @@ x101101011000000000101xxxxxxxxxx  n  cls     wx0 : wx5
 0x001110xx1xxxxx100001xxxxxxxxxx  n     add       dq0 : dq5 dq16 bhsd_sz
 0x001110xx1xxxxx100011xxxxxxxxxx  n     cmtst     dq0 : dq5 dq16 bhsd_sz
 0x001110xx1xxxxx100101xxxxxxxxxx  n     mla       dq0 : dq0 dq5 dq16 bhs_sz
-0x101111xxxxxxxx0000x0xxxxxxxxxx  n     mla       dq0 : dq5 dq16_idx_lhm bhsd_sz idx_lhm
+0x1011111xxxxxxx0000x0xxxxxxxxxx  n     mla       dq0 : dq5 dq16 vindex_SD sd_sz
+0x10111101xxxxxx0000x0xxxxxxxxxx  n     mla       dq0 : dq5 dq16_h_sz vindex_H h_sz
```

Comment:
So the bug being fixed was that the size was wrong for this variation?
How was the bug identified?  Could we capture how it was identified and make a test?

Comparing to external decoders does catch it just by disasm comparison:
```
$ disasm_a64 2f620820 
llvm-mc:   2f620820 mla v0.4h, v1.4h, v2.h[6]
capstone:  2f620820 mla v0.4h, v1.4h, v2.h[6]
bfd:       2f620820 mla v0.4h, v1.4h, v2.h[6]
DynamoRIO: 2f620820 mla %d1 $0x02 $0x01 $0x06 -> %d0
```

That's #4396.

Reply:
> So the bug being fixed was that the size was wrong for this variation?

No, it's a few things, the main one that the operand which should be a source register is an immediate int, e.g.
```
2f620820 : mla v0.4h, v1.4h, v2.h[6]                : mla    %d1 $0x02 $0x01 $0x06 -> %d0
2f620820 : mla v0.4h, v1.4h, v2.h[6]                : mla    %d1 %d2 $0x06 $0x01 -> %d0
                                                                 ^^^
```
The other problems with the existing implementation were:
- Use of new fields (`dq16_idx_lhm`, `idx_lhm`) incorrectly duplicating existing ones (`vindex_H`, `vindex_SD`, `dq16_h_sz`)
- Returning 2 bit integer operand for half-precision element indices rather than 3 bit in `decode_opnd_vindex_H()`.
  https://github.com/DynamoRIO/dynamorio/blob/d998d2bf371a53fe8c196c1152c75b5fa0c356ee/core/ir/aarch64/codec.c#L2389

> How was the bug identified?

I noticed the new fields when I returned from holiday but couldn't look at them properly until adding the new indexed vector `FMLS` and `FMUL` instructions.

> Could we capture how it was identified and make a test?

I think these problems would have been identified when adding IR creation tests in `suite/tests/api/ir_aarch64.c`, https://github.com/DynamoRIO/dynamorio/issues/2443, but we're currently focussing on decode rather than encode.

Where did you build `disasm_a64` from? I can't see it in the DR tree. Looks handy!

Reply:
```
$ type disasm_a64
disasm_a64 is a function
disasm_a64 () 
{ 
    bytes=`echo $1 | sed 's/\(..\)/0x\1 /g' | awk '{printf "%s %s %s %s\n", $4, $3, $2, $1}'`;
    echo "$bytes" | /usr/bin/llvm-mc -arch aarch64 --disassemble | grep --color=auto -v '\.text' | xargs echo "llvm-mc:   $1";
    echo "$bytes" | ~/bin/capstone -aarch64 - | cut -c 9-100 | xargs echo "capstone: ";
    echo -e ".global _start\n.align 6\n_start: .word 0x$1\n" > /tmp/_ddd.asm;
    aarch64-linux-gnu-as -g -o /tmp/_ddd.o /tmp/_ddd.asm || return;
    aarch64-linux-gnu-ld -static -o /tmp/_ddd /tmp/_ddd.o || return;
    aarch64-linux-gnu-objdump -D -d /tmp/_ddd | grep --color=auto -A 1 _start | grep --color=auto -v '_start>:' | cut -c 11-100 | xargs echo "bfd:      ";
    ~/dr/git/build_x64_tgt_a64_dbg/clients/bin64/drdisas ${bytes} | xargs echo "DynamoRIO:"
}
```

(For capstone I created a front-end program myself.  I can try to share it somewhere.)